File: 3437-maximum-total-damage-with-spell-casting/maximum-total-damage-with-spell-casting.py
class Solution:
    def maximumTotalDamage(self, power: List[int]) -> int:
        dp = defaultdict(int)
        for p in power:
            dp[p] += p
            dp[p - 3] = max(0, dp[p - 3])

        ans, last = 0, 0

        for p in sorted(dp.keys()):
            dp[p] = last = max(dp[p] + dp[p - 3], last)
            ans = max(ans, last)

        return ans

File: 3437-maximum-total-damage-with-spell-casting/maximum-total-damage-with-spell-casting.java
class Solution {
    public long maximumTotalDamage(int[] power){
        TreeMap<Integer, Long> dp = new TreeMap<>();
        for(int P : power){
            dp.put(P, dp.getOrDefault(P, 0L) + P);
            dp.put(P - 3, Math.max(0L, dp.getOrDefault(P - 3, 0L)));
        }
        
        long maxAns = 0 , last = 0;
        
        for(int P : dp.keySet()){
            dp.put(P, last = Math.max(dp.get(P) + dp.getOrDefault(P - 3, 0L), last));
            maxAns = Math.max(maxAns , last);
        }
        
        return maxAns;
    }
}